.github/workflows/unit-tests.yaml

```diff
@@ -0,0 +1,200 @@
+name: Unit Tests
+
+# Execute workflow for each PR and with each merge to the trunk
+on:
+  pull_request:
+  push:
+    branches:
+      - main
+
+# Cancel the workflow if any new changes pushed to a feature branch or the trunk
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: false # for now, false (put jobs in queue) allowing multiple PR testing consequently
+env:
+  VYPER_DOWNLOAD_URL: "https://github.com/vyperlang/vyper/releases/download"
+  VYPER_VERSION: "v0.3.10"
+  VYPER_DOWNLOAD_FILENAME: "vyper.0.3.10+commit.91361694"
+  CARGO_NET_GIT_FETCH_WITH_CLI: true
+  UNIT_TESTS_RESULTS_XML: unit-tests-results.xml
+  RUSTC_BOOTSTRAP: 1 # Use rustc in bootstrap mode allowing cargo2junit to use experimental --json feature
+
+jobs:
+  unit-tests-macos:
+    strategy:
+      matrix:
+        include:
+          - name: "MacOS x86"
+            runner: macos-12-large
+          - name: "MacOS arm64"
+            runner: [self-hosted, macOS, ARM64]
+    runs-on: ${{ matrix.runner }}
+    name: ${{ matrix.name }}
+    defaults:
+      run:
+        shell: zsh {0} # For MacOS hosts, .zprofile is pre-configured
+    env:
+      VYPER_DOWNLOAD_EXTENSION: ".darwin"
+    steps:
+      # TODO: This step should be done as part of github runners hooks
+      # Users should always expect clean workspace when a new job starts
```

Comment:
Not necessarily; I had some success with an unclean workspace at a previous employer.  (Admittedly an unusual case, with a complicated self-hosted Coverity installation.)  It might let us skip rebuilding LLVM unless necessary.  (For future consideration, not a blocker for this PR.)

Reply:
Yeah.

Although it can be useful in some cases, but I believe this is not the one. In my opinion for stability and parallel PR builds we should always start from the clean workspace. If we need to add some fixed dependencies, we should use cache or fixed runner directories, but not the workspace itself.